File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/PurchaseService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.endpoint.dto.PagedTicketsDto;
import at.ac.tuwien.sepm.groupphase.backend.entity.Ticket;

import java.util.List;

public interface PurchaseService {

    /**
     * Gets all purchased tickets for upcoming events of the currently logged-in user.
     *
     * @param userId the ID of the logged-in user
     * @return the purchased TicketOrder for upcoming events of the currently logged-in user
     */
    List<Ticket> getUpcomingPurchasedTickets(Long userId);

    /**
     * Gets all purchased tickets for past events of the currently logged-in user.
     *
     * @param userId the ID of the logged-in user
     * @return the purchased TicketOrder for past events of the currently logged-in user
     */
    PagedTicketsDto getPastPurchasedTickets(Long userId, int page, int size);

    /**
     * Cancels a purchased ticket.
     *
     * @param userId    the ID of the User that wants to cancel the Ticket
     * @param ticketIds the IDs of the Tickets to be cancelled
     */
    void cancel(Long userId, List<Long> ticketIds);
}

File: frontend/src/app/services/purchase.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Globals} from '../global/globals';
import {Observable} from 'rxjs';
import {PagedTicket, Ticket} from '../dtos/ticket';

@Injectable({
  providedIn: 'root'
})
export class PurchaseService {

  private readonly cartBaseUri: string = `${this.globals.backendUri}/purchases`;

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Get upcoming purchased Tickets of the currently logged-in user.
   */
  getUpcomingPurchasedTickets(): Observable<Ticket[]> {
    return this.httpClient.get<Ticket[]>(this.cartBaseUri + '/purchased/upcoming');
  }

  /**
   * Get past purchased Tickets of the currently logged-in user.
   */
  getPastPurchasedTickets(page: number = 0, size: number = 6): Observable<PagedTicket> {
    return this.httpClient.get<PagedTicket>(this.cartBaseUri + '/purchased/past' + '?page=' + page + '&size=' + size);
  }

  cancelPurchasedTickets(ticketIds: number[]) {
    console.log(`Cancel ticket with ids ${ticketIds}.`);
    return this.httpClient.put<Ticket[]>(this.cartBaseUri, {cancelTickets: ticketIds});
  }
}
